### ISSUE TITLE
Mikado prepared -> Invalid start and end values

### ISSUE THREAD

**Issue by francicco** [EI-CoreBioinformatics/mikado#317]
Hi,

I'm getting errors like this for the annotation generated with cufflinks.

```
Process AnnotationParser-3:
Traceback (most recent call last):
  File "/mnt/storage/home/tk19812/.local/lib/python3.7/site-packages/Mikado-2.0rc2-py3.7-linux-x86_64.egg/Mikado/parsers/gfannotation.py", line 93, in __init__
    self.start, self.end = tuple(fast_int(i, default=None) for i in self._fields[3:5])
  File "/mnt/storage/home/tk19812/.local/lib/python3.7/site-packages/Mikado-2.0rc2-py3.7-linux-x86_64.egg/Mikado/parsers/gfannotation.py", line 93, in <genexpr>
    self.start, self.end = tuple(fast_int(i, default=None) for i in self._fields[3:5])
SystemError: <built-in function fast_int> returned NULL without setting an error

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/mnt/storage/software/languages/anaconda/Anaconda3-2019-3.7/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/mnt/storage/home/tk19812/.local/lib/python3.7/site-packages/Mikado-2.0rc2-py3.7-linux-x86_64.egg/Mikado/preparation/annotation_parser.py", line 102, in run
    strand_specific=strand_specific)
  File "/mnt/storage/home/tk19812/.local/lib/python3.7/site-packages/Mikado-2.0rc2-py3.7-linux-x86_64.egg/Mikado/preparation/annotation_parser.py", line 435, in load_from_gff
    for row in gff_handle:
  File "/mnt/storage/home/tk19812/.local/lib/python3.7/site-packages/Mikado-2.0rc2-py3.7-linux-x86_64.egg/Mikado/parsers/GFF.py", line 368, in __next__
    line = GffLine(line)
  File "/mnt/storage/home/tk19812/.local/lib/python3.7/site-packages/Mikado-2.0rc2-py3.7-linux-x86_64.egg/Mikado/parsers/GFF.py", line 37, in __init__
    GFAnnotation.__init__(self, line, my_line, header=header)
  File "/mnt/storage/home/tk19812/.local/lib/python3.7/site-packages/Mikado-2.0rc2-py3.7-linux-x86_64.egg/Mikado/parsers/gfannotation.py", line 95, in __init__
    raise ValueError("Invalid start and end values: {}".format(" ".join(self._fields[3:5])))
ValueError: Invalid start and end values: 7506360 7506699
Mikado prepared has encountered a fatal error. Please check the logs and, if there is a bug,report it to https://github.com/EI-CoreBioinformatics/mikado/issues
```

I don't understand what's causing that.
Thanks
Francesco

**Comment by francicco** [comment]
Hi @lucventurini,

This happens mainly with cufflinks outputs,
Still haven't figured out why those positions are problematic.
I'm getting mad!

Cheers
Francesco

**Comment by francicco** [comment]
Hi @lucventurini,

Any help for this?
Thanks
Francesco

**Comment by lucventurini** [comment]
Hi @francicco

I will have a deep look at it once I am back from holidays, on the 3rd.

@ljyanesm, if you have time, would you be able to check this?

Kind regards

Luca Venturini

**Comment by francicco** [comment]
It's ok, I went through for this project, but I can totally use it for the next few genomes to annotate.
Thanks
F

**Comment by lucventurini** [comment]
Dear @francicco 

I am now back from holidays. Would you please be able to send me a snippet from the Cufflinks file that is causing the problem? I am investigating.

I think that the issue is that I am using a stable but non-standard library for converting strings into integers, [fastnumbers](https://github.com/SethMMorton/fastnumbers). It does seem to croak on this input.

It would help also to know about the version of fastnumbers you have installed in your environment:

```python
$ python
>>> import fastnumbers
>>> fastnumbers.__version__
'3.0.0'
```

After this I can try to get to the bottom of the bug.

Many thanks

**Comment by francicco** [comment]
Hi Luca,

this is one of the annotation that fails

```
Eisa2100	Cufflinks	mRNA	6552421	6566205	1019	+	.	ID=TrG9896.TU15265;Parent=TrG9896
Eisa2100	Cufflinks	exon	6552421	6552509	1019	+	.	ID=exon1_TrG9896.TU15265;Parent=TrG9896.TU15265
Eisa2100	Cufflinks	CDS	6552422	6552509	1019	+	0	ID=cds_of_TrG9896.TU15265;Parent=TrG9896.TU15265
Eisa2100	Cufflinks	exon	6562518	6562620	1019	+	.	ID=exon2_TrG9896.TU15265;Parent=TrG9896.TU15265
Eisa2100	Cufflinks	CDS	6562518	6562620	1019	+	2	ID=cds_of_TrG9896.TU15265;Parent=TrG9896.TU15265
Eisa2100	Cufflinks	exon	6564235	6564426	1019	+	.	ID=exon3_TrG9896.TU15265;Parent=TrG9896.TU15265
Eisa2100	Cufflinks	CDS	6564235	6564426	1019	+	1	ID=cds_of_TrG9896.TU15265;Parent=TrG9896.TU15265
Eisa2100	Cufflinks	exon	6564832	6565084	1019	+	.	ID=exon4_TrG9896.TU15265;Parent=TrG9896.TU15265
Eisa2100	Cufflinks	CDS	6564832	6565084	1019	+	1	ID=cds_of_TrG9896.TU15265;Parent=TrG9896.TU15265
Eisa2100	Cufflinks	exon	6565824	6566205	1019	+	.	ID=exon5_TrG9896.TU15265;Parent=TrG9896.TU15265
Eisa2100	Cufflinks	CDS	6565824	6566183	1019	+	1	ID=cds_of_TrG9896.TU15265;Parent=TrG9896.TU15265
```

my version is `fastnumbers-2.2.1`

F

**Comment by lucventurini** [comment]
Dear @francicco 

Many thanks, I'll try as soon as possible. As a quick complementary check, if you could upgrade `fastnumbers` to 3.0.0 and see whether that solves the problem on your end, we would be grateful.

Due to the bugs you reported, @ljyanesm and I are anyway considering whether to remove `fastnumbers` for good.

**Comment by francicco** [comment]
I'll reinstall the new release (is it with python3 right)
F

**Comment by francicco** [comment]
Updated with the new version and `fastnumbers==3.0.0`. Same error.
F

**Comment by lucventurini** [comment]
Dear @francicco 

I just tried to parse the snippet you sent and I could not reproduce the error - the parser breezes through.

My hunch (hope maybe?) is therefore that something must be invisibly garbled in those lines, maybe an extra invisible character that throws `fastnumbers`. Could you grep these lines from the GTF, please, redirect them to a new snippet file, and attach the file here as a TXT? This is because copying/pasting might have removed the problem.

Another possibility is that fastnumbers is somehow malfunctioning on your specific system. If you have installed it via pip, may I ask you to uninstall it and reinstall `fastnumbers` [via conda](https://anaconda.org/conda-forge/fastnumbers) in the `mikado` environment?

A last resort, which could either fix the bug or tell us better what is going on, is to modify this file:

```
/mnt/storage/home/tk19812/.local/lib/python3.7/site-packages/Mikado-2.0rc2-py3.7-linux-x86_64.egg/Mikado/parsers/gfannotation.py
```

at line **93**, from

```python
            self.start, self.end = tuple(fast_int(i, default=None) for i in self._fields[3:5])
```

to

```python
            self.start, self.end = [int(i) for i in self._fields[3:5])
```

At a minimum, the python built-in `int` function should fail more informatively.

Many apologies for so much fiddling, but this bug seems to be quite subtle in nature. Thank you for your patience and help in teasing it out!

Kind regards

Luca Venturini

**Comment by lucventurini** [comment]
Closed in  67dd3dc.
